resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsEventHandler.scala

```diff
@@ -0,0 +1,229 @@
+            // deletion event followed by some arbitrary modification event - we want the
+            // deletion event to "stick".
+            if (pendingExecutors.contains(execId)) {
+              pendingExecutors.remove(execId)
+              runningExecutors.add(execId)
+            }
+          // TODO (SPARK-24135) - handle more classes of errors
+          case "error" | "failed" | "succeeded" =>
+            // If deletion failed on a previous try, we can try again if resync informs us the pod
+            // is still around.
+            // Delete as best attempt - duplicate deletes will throw an exception but the end state
+            // of getting rid of the pod is what matters.
+            if (!isDeleted(updatedPod)) {
+              Utils.tryLogNonFatalError {
+                kubernetesClient
+                  .pods()
+                  .withName(updatedPod.getMetadata.getName)
+                  .delete()
+              }
+            }
+            removeExecutorFromSpark(schedulerBackend, updatedPod, execId)
+        }
+      }
+    }
+
+    val currentRunningExecutors = runningExecutors.size
+    val currentTotalExpectedExecutors = totalExpectedExecutors.get
+    if (pendingExecutors.isEmpty && currentRunningExecutors < currentTotalExpectedExecutors) {
+      val numExecutorsToAllocate = math.min(
+        currentTotalExpectedExecutors - currentRunningExecutors, podAllocationSize)
+      logInfo(s"Going to request $numExecutorsToAllocate executors from Kubernetes.")
+      val newExecutorIds = mutable.Buffer.empty[Long]
+      val podsToAllocate = mutable.Buffer.empty[Pod]
+      for ( _ <- 0 until numExecutorsToAllocate) {
+        val newExecutorId = EXECUTOR_ID_COUNTER.incrementAndGet()
+        val executorConf = KubernetesConf.createExecutorConf(
+          conf,
+          newExecutorId.toString,
+          applicationId,
+          driverPod)
+        val executorPod = executorBuilder.buildFromFeatures(executorConf)
+        val podWithAttachedContainer = new PodBuilder(executorPod.pod)
+          .editOrNewSpec()
+          .addToContainers(executorPod.container)
+          .endSpec()
+          .build()
+        kubernetesClient.pods().create(podWithAttachedContainer)
+        pendingExecutors += newExecutorId
+      }
+    } else if (currentRunningExecutors == currentTotalExpectedExecutors) {
+      logDebug("Current number of running executors is equal to the number of requested" +
+        " executors. Not scaling up further.")
+    } else if (pendingExecutors.nonEmpty) {
+      logInfo(s"Still waiting for ${pendingExecutors.size} executors to begin running before" +
+        s" requesting for more executors.")
+    }
+  }
+
+  def sendUpdatedPodMetadata(updatedPod: Pod): Unit = {
+    eventQueue.add(Seq(updatedPod))
+  }
+
+  def sendUpdatedPodMetadata(updatedPods: Iterable[Pod]): Unit = {
+    eventQueue.add(updatedPods.toSeq)
```

Comment:
Instead of storing a `Seq` of events, how about storing events directly, and calling `eventQueue.add` for each pod here? Then you don't need to call `flatten` when processing the events.

Reply:
Thought about this, I prefer the idea of the polling executor being able to add all of its pods that it saw in a round as an atomic operation. But as long as we're handling out of order events properly perhaps the difference is negligible.